File: F#/Shared/Gravatar.fs
module Gravatar

open System
open Helpers

#if __IOS__
open MonoTouch.Foundation
#endif

type Rating =
| G | PG | R | X

let private ratingToString = function
    | G -> "g"
    | PG -> "pg"
    | R -> "r"
    | X -> "x"

let private url = "http://www.gravatar.com/avatar.php?gravatar_id="

let private MD5Hash (input : string) =
    System.Security.Cryptography.MD5.Create().ComputeHash (System.Text.Encoding.Default.GetBytes (input))
    |> Array.map (fun x->x.ToString("x2"))
    |> String.concat ""

let private GetUrl (email:string) size rating =
    let hash = MD5Hash (email.ToLower ())

    if size < 1 || size > 600 
    then Failure "The image size should be between 20 and 80"
    else Success (url + hash + "&s=" + size.ToString () + "&r=" + (ratingToString rating))
        
let GetImageBytes email size rating = async {
    match GetUrl email size rating with
    | Failure message -> return [||]
    | Success url -> let client = new System.Net.WebClient ()
                     return! (client.DownloadDataTaskAsync url |> Async.AwaitTask) }

#if __IOS__
let GetImageData email size rating = async {
    let! data = GetImageBytes email size rating
    return NSData.FromStream(new System.IO.MemoryStream(data)) }
#endif

File: Shared/Data/Gravatar.cs
using System;
using System.IO;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using System.Security.Cryptography;

#if __IOS__
using MonoTouch.Foundation;
#endif

namespace XamarinStore
{
	class Gravatar
	{
		public enum Rating { G, PG, R, X }

		const string _url = "http://www.gravatar.com/avatar.php?gravatar_id=";

		public static string GetURL (string email, int size, Rating rating = Rating.PG)
		{
			var hash = MD5Hash (email.ToLower ());

			if (size < 1 | size > 600) {
				throw new ArgumentOutOfRangeException("size", "The image size should be between 20 and 80");
			}

			return _url + hash + "&s=" + size.ToString () + "&r=" + rating.ToString ().ToLower ();
		}

		public static async Task<byte[]> GetImageBytes (string email, int size, Rating rating = Rating.PG)
		{
			var url = GetURL (email, size, rating);
			var client = new WebClient ();
			return await client.DownloadDataTaskAsync (url);
		}

#if __IOS__
		public static async Task<NSData> GetImageData (string email, int size, Rating rating = Rating.PG)
		{
			byte[] data = await GetImageBytes (email, size, rating);
			return NSData.FromStream (new System.IO.MemoryStream (data));
		}
#endif

		static string MD5Hash (string input)
		{
			var hasher = MD5.Create ();
			var builder = new StringBuilder ();
			byte[] data = hasher.ComputeHash (Encoding.Default.GetBytes (input));

			foreach (byte datum in data)
				builder.Append (datum.ToString ("x2"));

			return builder.ToString ();
		}
	}
}

